### PR TITLE
refactor: update config for immutability and extension

### PR DESCRIPTION
**Pull request by aaron-steinfeld**
This change is to support adding further configuration in the form of attribute client config. When adding this, the use of ConfigBeanFactory prevented our standard dotted notation (because each top level property must match a property in the pojo). Digging in further, there were a number of antipatterns with the bean factory and existing config parsing- for example, the coupling of method names and config, the distribution of parsing logic across the codebase, the mutability of configs and confusing naming.

This change takes a cut at cleaning this up in code, while leaving the config format intact to ensure compatibility with existing configs.

### CODE REVIEW

File: query-service-impl/src/test/resources/application.conf [hunk 1]
```diff
@@ -2,6 +2,12 @@ service.name = "query-service"
 service.port = 8090
 service.admin.port = 8091
 service.config = {
+  attribute.client = {
```
Comment (buchi-busireddy) [inline query-service-impl/src/test/resources/application.conf:5]: Where are you using these? I don't see the code to validate attributes yet.

File: query-service-impl/src/main/java/org/hypertrace/core/query/service/QueryServiceConfig.java [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+    private static final String CONFIG_PATH_CLIENT_KEY = "clientConfig";
+    private static final String CONFIG_PATH_REQUEST_HANDLER_INFO = "requestHandlerInfo";
+
+    private final String name;
+    private final String type;
+    private final String clientConfig;
+    private final Config requestHandlerInfo;
+
+    private RequestHandlerConfig(Config config) {
+      this.name = config.getString(CONFIG_PATH_NAME);
```
Comment (buchi-busireddy) [inline query-service-impl/src/main/java/org/hypertrace/core/query/service/QueryServiceConfig.java:55]: Just for my own understanding, what's the difference b/w this model and the previous one like:
`ConfigBeanFactory.create(config, QueryServiceImplConfig.class);`?

**Review by buchi-busireddy — APPROVED** [review]

File: query-service-impl/src/test/resources/application.conf (hunk 1 above)
Comment (aaron-steinfeld) [reply query-service-impl/src/test/resources/application.conf:5]: These actually aren't used yet, they're used in the following change for projections - that change drove this one, so I split this part off to shrink the review. Should have probably removed these.

File: query-service-impl/src/main/java/org/hypertrace/core/query/service/QueryServiceConfig.java (hunk 2 above)
Comment (aaron-steinfeld) [reply query-service-impl/src/main/java/org/hypertrace/core/query/service/QueryServiceConfig.java:55]: Discussed in the PR description - https://github.com/hypertrace/query-service/pull/23#issue-491929367 
Specifically, the problem that drove this was the fact that every top level property in the bean object had to correspond to a top level config property.

**Review by tim-mwangi — APPROVED** [review]

**Pull request merged**